server/src/main/java/org/opensearch/index/shard/RemoteStoreRefreshListener.java

```diff
@@ -209,15 +209,15 @@ boolean uploadNewSegments(Collection<String> localFiles) throws IOException {
                 );
                 return true;
             }
-        }).forEach(file -> {
-            try {
-                remoteDirectory.copyFrom(storeDirectory, file, file, IOContext.DEFAULT);
-            } catch (IOException e) {
-                uploadSuccess.set(false);
-                // ToDO: Handle transient and permanent un-availability of the remote store (GitHub #3397)
-                logger.warn(() -> new ParameterizedMessage("Exception while uploading file {} to the remote segment store", file), e);
-            }
-        });
+        }).collect(Collectors.toList());
+
+        try {
+            remoteDirectory.copyFilesFrom(storeDirectory, filteredFiles, IOContext.DEFAULT);
+        } catch (Exception e) {
```

Comment:
There can be other future exceptions raised here as well. We are anyways not doing anything if uploadStatus is false in `RemoteStoreRefreshListener` apart from printing a log in throwable catch block.

Reply:
There can be other future exceptions raised here as well. We are anyways not doing anything if uploadStatus is false in `RemoteStoreRefreshListener` apart from printing a log in throwable catch block.

Reply:
If we do serial uploads, then we will end up having huge refresh latencies which is what is happening in the old flow. In my opinion, we shouldn't compromise on performance unless it beats correctness. 
Every file upload has an associated future. For stats use case, we should be able update the status in the respective future callback.

Reply:
shall we attempt parallelising from one layer above to make sure we are able to add file specific tracking? We should be able to refactor the code to achieve the same. Else events like segment merges or in general flushes would appear binary and also would lead to insufficient data for the user for stats, operation and maintainability.

Reply:
Currently, we don't do any sort of upload management in `RemoteStoreRefreshListener`. If we move it up, then transfer code will seem sprinkled across multiple classes. Higher layer knows which files are being submitted for transfer and we can easily attach callbacks/listeners when transfer reaches a terminal state. Do you see a concern in this approach?
My plan is not only to make segment transfers async but the entire callback flow async including metadata upload and notification flow. Migration of segment transfers to async IO is first step. With async IO we don't need any thread to be unnecessarily occupied and waiting for response and therefore, the entire callback flow should ideally work on async model. So, with this plan we anyways need respective file level callbacks attached and code will no longer remain sequential. Therefore, even if we move the parallel code up today, we will anyways have to migrate it away tomorrow and that's the reason I am leaning towards callback model for stats and other things as well.

Reply:
I've rebased latest changes and made a change in `RemoteSegmentStoreDirectory#copyFilesFrom`, which now takes an `UploadTracker` as parameter and internally calls relevant hooks during the upload process
```
public boolean copyFilesFrom(Directory from, Collection<String> files, IOContext context, UploadTracker uploadTracker)
```